src/Bicep.Core/Analyzers/Linter/Rules/NoHardcodedEnvironmentUrlsRule.cs

```diff
@@ -10,89 +10,121 @@
-            private readonly Regex exclusionRegex;
+            private readonly ImmutableArray<string> excludedHosts;
 
-            public Visitor(Regex disallowedRegex, int minHostLen, Regex exclusionRegex)
+            public Visitor(ImmutableArray<string> disallowedHosts, int minHostLen, ImmutableArray<string> excludedHosts)
             {
-                this.hostRegex = disallowedRegex;
+                this.disallowedHosts = disallowedHosts;
                 this.minHostLen = minHostLen;
-                this.exclusionRegex = exclusionRegex;
+                this.excludedHosts = excludedHosts;
+            }
+
+            public static IEnumerable<(TextSpan RelativeSpan, string Value)> FindHostnameMatches(string hostname, string srcText)
+            {
+                bool isExactDomainMatch(int index)
+                {
+                    var leadingAlnum = index > 0 && 
+                        char.IsLetterOrDigit(srcText[index - 1]);
+                    var trailingAlnum = index + hostname.Length < srcText.Length && 
+                        char.IsLetterOrDigit(srcText[index + hostname.Length]);
+
+                    return !leadingAlnum && !trailingAlnum;
+                }
+
+                if (hostname.Length == 0)
+                {
+                    // ensure we terminate - the below for-loop uses this value to increment
+                    yield break;
+                }
+
+                var matchIndex = -1;
+                for (var startIndex = 0; startIndex <= srcText.Length - hostname.Length; startIndex = matchIndex + hostname.Length)
+                {
+                    matchIndex = srcText.IndexOf(hostname, startIndex, StringComparison.OrdinalIgnoreCase);
+                    if (matchIndex < 0)
+                    {
+                        // we haven't foud any instances of the hostname
+                        yield break;
+                    }
+
+                    // check preceeding and trailing chars to verify we're not dealing with a substring
+                    if (isExactDomainMatch(matchIndex))
+                    {
+                        var matchText = srcText.Substring(matchIndex, hostname.Length);
+                        yield return (RelativeSpan: new TextSpan(matchIndex, hostname.Length), Value: matchText);
+                    }
+                }
+            }
+
+            public IEnumerable<(TextSpan RelativeSpan, string Value)> RemoveOverlapping(IEnumerable<(TextSpan RelativeSpan, string Value)> matches)
+            {
+                TextSpan? prevSpan = null;
+                foreach (var match in matches.OrderBy(x => x.RelativeSpan.Position).ThenByDescending(x => x.RelativeSpan.Length))
+                {
+                    if (prevSpan is not null && TextSpan.AreOverlapping(match.RelativeSpan, prevSpan))
+                    {
+                        continue;
+                    }
+
+                    yield return match;
```

Comment:
>yield [](http://example.com/codeflow?start=20&length=5)

Should we add a comment that this rule is on the critical path?

Also, I'm guessing you went with `yield` to avoid list allocations? Worth commenting on those types?